### ISSUE TITLE
Where to upload assets repositories?

### ISSUE THREAD

**Issue by illwieckz** [Unvanquished/Unvanquished#979]
I'm working on asset repositories to make unvanquished packages reproduceables and to be able to track the changes and ensure we don't lose any lossless original (and this will also allows mappers to build against lossless textures, as a side effect).

It's about textures repositories, sound repositories, etc. etc.

Where can I upload them? Do I have to upload them under the @Unvanquished organization, or is it better to create something like an _UnvanquishedArtists_ organization for asset-only repositories ?

**Comment by Viech** [comment]
If on GitHub, @Unvanquished is already pretty big (disk space wise), so @UnvanquishedArtists would be preferable in this case.

I'm not sure if git is ideal due to its history tracking that might let the repository explode if (small) changes are made to the files. Maybe one can limit it to a certain amount of tracked changes per file or purge the history manually from time to time?

**Comment by DolceTriade** [comment]
I don't think Github has organization limits, only repo limits.

**Comment by illwieckz** [comment]
Yes I was suggesting creating a dedicated asset organization not because file size but because places being messy.

Git is not fine with large assets repositories yes, but I'm not fine myself with one big repository to rules them all, whatever the underlying technology. Xonotic is doing this and you can't download a map or a texture set without cloning all the datas and it's boring. Even if Git were able to split binaries in chunk or dealing will rolling checksums like bup does, I will not be fine with a big repository to rules them all. Even if Git were able to clone a subdir of the a repository I would say the same.

And well, there is now pakpath options in both dæmon and q3map2, people can put their asseits where they want…

I clearly prefer one parent repository with one submodule repository per texture package, map etc. So my question for the organization is more about messy places.

The idea is also allowing people to clone a texture repository so they can build their lightmaps against the lossless ones for example, same with map models. This way People can clone only what they need, for example the pk02 one. It means we can also create an empty parent directory with all assets as submodules (textures, maps, sounds, models, etc.) but also a parent repositories with all textures repositories as submodules, why not.

In end there will be one repository per texture package, one per map etc., and perhaps 3 models repositories : buildables, players, weapons because these ones are the larger ones and are expected to change a lot. About the sounds, my idea is to make a legacy package for the next release, and create a new sound repository for the newer ones, the newer package overriding the legacy one in DEPS priority: it means the legacy one will be left untouched, and will be dropped when all sounds will be replaced. Textures, maps, sounds, models, it means many repositories.

My current work in progress repositories allows anyone to build one package or to build them all from a parent dir.

If we put asset repositories at @Unvanquished it will bebome messy as hell, and I'm not sure it's a good idea. Also, we can define some different contribution/merge rules for asset repositories (because it's not code!).

The idea of an @UnvanquishedArtists organization is also to not mess artists with code repositories, the @UnvanquishedArtists organization acting as a directory for asset repositories only.

If we have to upload asset managing tools, the place for these tools would be in non-asset @Unvanquished place, or in @DaemonDevelopers place if it's generic enough to be used by third-party projects.

**Comment by Viech** [comment]
> I clearly prefer one parent repository with one submodule repository per texture package, map etc.

I like the idea, if it's not too fine grained (i.e. too many repositories). However, I think that one repo for weapons, buildables and player models each is actually not fine grained enough. This is something we should determine in advance. With a good selection of subrepos, your idea allows us to completely purge history for selected assets, if the subrepos grow too big for no good reason!

> If we put asset repositories at @Unvanquished it will bebome messy as hell, and I'm not sure it's a good idea.

I agree. I think `UnvanquishedAssets` would be a good organizationb name (since the main one is just called `Unvanquished` and not `UnvanquishedProgrammers` or so), maybe with `UnvRoot` as the root repo and `UnvDragoon` and names like that for subrepos.

I also don't expect our artists to actually get into git and learn about the pitfalls of having a binary repository, so it will mostly be programmers maintining the asset repositories. It also automatically adds another layer of quality control/review, so that's a good thing.

> About the sounds, my idea is to make a legacy package for the next release, and create a new sound repository for the newer ones, the newer package overriding the legacy one in DEPS priority: it means the legacy one will be left untouched, and will be dropped when all sounds will be replaced.

Sounds like a good approach. So they will like this one! 😜

**Comment by illwieckz** [comment]
> However, I think that one repo for weapons, buildables and player models each is actually not fine grained enough.

Yes, and I already splited my model stuff in 3 repos: buildables, players and models. juste because of the repo size!

> I agree. I think UnvanquishedAssets would be a good organization name

That's OK to me, I was suggesting @UnvanquishedArtists to mimic @DaemonDevelopers but I'm OK with UnvanquishedAssets (that was the name was thinking about first).

For subrepos I guess the better is just to name them as pk3dir, even with extension etc. this way it's explicit it's a source pk3dir, like `tex-space_source.pk3dir`. You just clone it et voilà !

> I also don't expect our artists to actually get into git

Yes I expect artists just send stuff like they already do. It's not a problem at all, and it helps to take care a lot of the repository (specially his size).

About assets management and problems related to git. It's doable if we follow strict rules, for example never comitting directly to master and avoiding foreign branch merge directly on GitHub except pre-release branches that are known to be clean and tight. The admin has all power to make things better before merging master.

Also, we can keep in mind some repositories are not expected to change a lot, like texture ones. Map repositories can change a lot but if they rely properly on texture repositories it's not a problem, and the .map file is a text file which is nicely diffable if we denumberize entities before committing (something the maintainer can do retroactively on development branch before merging to master).

In many case the artist will just send the assets as we already do, so the asset maintainer just includes them in the pre-release branch. If an artist knows how to use git he can submit a PR, this way the maintainer fetches the artist branch and rewrite the tree if needed.

I wrote a script that can rewrite a git tree converting assets from previous commits (for example from .tga to .png), and I wrote a script that can rewrite a git tree to recompress assets: zopflying the .png, bzip2ing the .xcf, gziping the .blend (gimp and blender can open bzip2ed .xcf and gziped .blend without modification), flacing .wav (they are meant to be opused at build time, so the care is about user's editor format, not engine format). To make history looking better, as said before it's easy to write a script to denumberize .map files in history or fixing line ending in scripts or removing useless texture extension in map files and shader files. The maintainer can clean all the stuff before merging, ensuring everything is clean and tighted compressed with best algorithm available before putting stuff into master.

Since it's assets and it's different than code, we cans safely say the pre-release branch can be force-pushed at anytime. Since assets are mainly binary stuff, there is no need to care about diffs. Thanks to our FS organisation it's very easy to exchange commits between two foreign trees since they all shares the exact same tree. The git repository itself is there to track what was done and to not lose things, it's not expected to be the primary place to exchange stuff.

The idea is to rewrite the history before merging on master and saying clearly that every stuff expected to be merged can and will be rewritten (first need: optimal compression), because it's an asset repository so the rules change. And we must avoid reverts in any case, preferring rewriting the history if needed, before merging. I guess the pre-release branch must not be merged to master until the very last time to allow rewriting stuff until the very last time. If a file have seen multiple changes before the last release (last merge to master), perhaps only the last version must be kept in the history to be merged, to save space, why not. This way we will be able to keep the master branch always tight without having to rewrite it, since the rewriting must be avoided if possible. I guess it's fair if we clearly announce the pre-releases branches are meant to be force pushed by the maintainers.

The needs for asset repositories are

- taking care of originals for assets we are using
- having a change history with enough metadata like commiter/date/comment
- being able to tag versions
- being able to pick a revision if needed
- being able to reproduce the asset build
- allowing people to build against originals
- being able to build prerelease stuff, even assets

Even if we have to rewrite some history one day (worst thing to do on a repo, but it can be discussed at some milestone for example if an awesome wide supported lossless format divides significantly the whole repo size), all these previous needs will be still fullfilled and it will still be a huge improvement.

Having assets in a git history means we will be able to numberize the package from a git tag, or be able to build a package that only contains modified stuff since the previous tag. Automagically. I'm planning to extend my toolbox to do that: packaging a pk3 containing only modified stuff since an arbitrary tag, then versionning the pk3 on git metadatas (the DEPS file can be automagically written with the right previous tag as a dependency!)

With a git history, a package build tool can add itself the version based on `tag`, and if the last commit is not tagged, it can use something like `tag+last-commit-date~commit-id` as version number.

@Viech can you create that assets organisation? :smiley:

**Comment by Viech** [comment]
Let's please keep our posts here much shorter than what you just did, atleast provide a TL;DR, people might be detered from entering the discussion otherwise. 😉 

I'll create the organization later today, right now I'm at work.

I would be fine with using the proper source pk3dir names for the sub-repositories as you just proposed, but keep in mind that **the asset repository needs to host source files that we do not ship at all with the game** (think of .blend and .xcf files), and the build scripts that create release pk3s from the sub-repositories will need to identify and ignore these files in some way.

@illwieckz, in case we do go with such pk3dirs as sub-repositories, can you figure out a full list of the pk3dirs you would create and post it here first? As we will ship the very same pk3s with the game, we should discuss this briefly.

I think it would be good if we had the pk3 build scripts inside the root repository, so that one could actually generate all of Unvanquished's pk3s just by cloning and "compiling" the root repository. The root repository could have a `source` folder where the sub-repositories are. The build script would then generate a `pkg` folder where it puts the release/patch pk3s. Symlinking that `pkg` folder from within Unvanquished's build directory would then allow you to start the game with your own copy of the pk3s.

Minor: I changed my mind about the root repository's name, I think we should call it `UnvanquishedAssets` just like the organization.

**Comment by Viech** [comment]
@illwieckz, I have created @UnvanquishedAssets and invited you. All organization members can add repositories and add *outside collaborators* to them. @DolceTriade, @kharnov, if have invited you as owners. You can add more organization members according to the following rules:
- Only maintainers of our entire asset base or developers of the associated toolchain (that generates our release pk3s from the asset source repositories) should be added as members.
- For people who maintain individual assets we shall try to grant them *outside collaborator* status to the affected repositories first and see if this is sufficient for their workflow.

**Comment by illwieckz** [comment]
TLDR:

Everything was thought this way. I'm happy because we share the same ideas. In fact I'm just in need for that asset organization because publishing some preview repositories as example will answer 95% of the questions.

_____________

> Let's please keep our posts here much shorter than what you just did, atleast provide a TL;DR, people might be detered from entering the discussion otherwise.

Yes Yes I'm a writing man. In fact I already split my initial writing over the time and some parts were never published there. It would have been _worst_. :scream:

> but keep in mind that the asset repository needs to host source files that we do not ship at all with the game

That's currently what I'm doing, but the first need of these repos are to be able to build the packages, not necessarily to redo the iqm from the blend or some textures from models. The build targets of these repositories are `pk3`, not `model.iqm`. I really hope to be able to store all needed stuff in these repositories: `.blend` etc. but some issues raises with Photoshop.

In the asset archive kharnov sent me there is some PSD for diffuse textures that weight between between 100 and 400M each one. :scream: If I convert them to bzip2 xcf I save a lot of space, but I fear to lose something in the process so I haven't done nothing with them. I want to put everything in these repositories but what put limits in our imagination is not Git, it's Photoshop. There is also the question for assets contributed we are not using yet or we are not using anymore. Perhaps it would be a special place for that.

I'm currently discussing the source repositories of the `pk3`. There is some questions pending for "the sources of the sources", and I refrain to put a _Photoshop rant_ there because this comment is already too long. Well, I'm writing another wall of text but you will find 90 of your answers in the way I designed my pk3s, just publishing some previews will answer many questions.

> and the build scripts that create release pk3s from the sub-repositories will need to identify and ignore these files in some way.

The build tool I'm writing already recognizes file "source of source" (xcf, blend etc.) to ignore them without configuration, it also handles ignore lists we can manually edit (in fact it's a task list, you can ask to package a file that would be ignored by default, or to ignore as file that would be packaged by default). This question was solved years _before_ I begin the current repository work we are talking about.

> can you figure out a full list of the pk3dirs you would create and post it here first?

Yes, it's something I haven't done in previous comments because they were already too long. :wink: 

> I think it would be good if we had the pk3 build scripts inside the root repository

That's how everything was thought.

> Minor: I changed my mind about the root repository's name, I think we should call it UnvanquishedAssets just like the organization.

That's good, we will be able to do something like that:

```
# source pk3dir (png, etc. in pk3dir)
q3map2 -fs_pakpath …/UnvanquishedAssets/src […]

# test pk3dir (crn, etc. in pk3dir)
daemon -pakpath …/UnvanquishedAssets/build/test

# final pk3 (crn, etc. in versioned pkg)
daemon -pakpath …/UnvanquishedAssets/build/pk3
```

**Comment by Viech** [comment]
> the first need of these repos are to be able to build the packages, not necessarily to redo the iqm from the blend or some textures from models.

I do not ask for re-compiling source files like that. (With *sources* I mean stuff like .xcf files. I would call a .png file an *uncompressed* asset and a .crn file a *compressed* one. No need to call sources the "source of a source".) The important thing is that we *store* our sources at a convenient place. The same goes for license files. Every sub-repository needs to contain one that lists all authors and licenses relevant to the asset(s) in question. Unlike sources, these need to be shipped with the game, though.

> In the asset archive kharnov sent me there is some PSD for diffuse textures that weight between between 100 and 400M each one.

I hear GitHub only has per-repository file size limits, not per-organization. But a few hundred megabytes is still a lot if someone wants to clone the whole thing. We may want to consider using an external host for these files and just link them from within the repository.

---

If you have any trouble moving my map texture package toolchain to the asset host hook me up. It should be doable with `sloth` as another subrepo I guess. Luckily I did put the uncompressed textures in pk3dirs, so your dream of compiling maps against uncompressed ones should be in reach! I just hope my shell scripts are portable…

**Comment by illwieckz** [comment]
When I said “or some textures from models” I was meaning redesigning texture from modeling tools, not tweaking them in XCF or similar software.

> We may want to consider using an external host for these files and just link them from within the repository.

Hopefully there is only a couple of them, and it's mainly an issue related to the way Adobe Photoshop writes the PSD format. If someone owns a genuine Photoshop and is used to it, it could be nice if he can rewrite them from Photoshop using some tweaks, or, far better, check if there is nothing list when converted to XQF, and this case, we would be able keep the converted and compressed XQF instead.

> If you have any trouble moving my map texture package […]

I've experienced no one issue with your asset folder, thanks. I already played some games using packages built against my wip repositories without noticing texture issues.

**Comment by Viech** [comment]
> I've experienced no one issue with your asset folder, thanks.

I was also refering to the shell scripts of my pk3-packaging toolchain. I assume you aim to incorporate/rewrite them in(to) your own build-all-release-pk3s toolchain?

**Comment by illwieckz** [comment]
assets repositories are there: https://github.com/UnvanquishedAssets/UnvanquishedAssets
